### PR TITLE
feat(cosmos): opportunities metadata

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Adds a resolver for Cosmos opportunities metadata - not handling the default opportunity case yet.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

None, under flag and not consumed in the view layer yet

### Engineering

- Connect a wallet with an active validator (default validators aren't fetched yet)
- Enable `CosmosSdkOpportunitiesAbstraction`
- Ensure the validator metadata is matching the one in the current `validatorData` slice / view-layer data

### Operations

None

## Screenshots (if applicable)

image
image
image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3736** <a href="https://app.graphite.dev/github/pr/shapeshift/web/3736" target="_blank">Graphite</a> 
    * **PR #3741** <a href="https://app.graphite.dev/github/pr/shapeshift/web/3741" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/3741?utm_source=stack-comment).

File: src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts [hunk 1]
```diff
@@ -60,3 +70,91 @@ export const cosmosSdkOpportunityIdsResolver = async ({
...
+}: OpportunitiesMetadataResolverInput): Promise<{
+  data: GetOpportunityMetadataOutput
+}> => {
+  const stakingOpportunitiesById: Record<StakingId, OpportunityMetadata> = {}
+
+  const state = reduxApi.getState() as ReduxState
+
+  const { CosmosSdkOpportunitiesAbstraction } = selectFeatureFlags(state)
+
+  // TODO(gomes): default validatorIds
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts:87]: Implement me before opening

- No ValidatorIds length -> use both default validators for Cosmos and Osmosis
- Has length -> check if there is an ATOM/OSMO AccountId

Both can be abstracted as two `hasCosmosValidatorId` and `hasOsmosisValidatorId`, `false` for both for the first case, `boolean` for the latter

File: src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts [hunk 2]
```diff
@@ -60,3 +70,91 @@ export const cosmosSdkOpportunityIdsResolver = async ({
...
+        if (isRejected(settledValidator)) {
+          moduleLogger.error(settledValidator.reason, 'Error fetching Cosmos SDK validator')
+        }
+        if (isFulfilled(settledValidator) && settledValidator.value) {
+          const { validatorId, ...opportunityMetadata } = settledValidator.value
+
+          acc[validatorId] = opportunityMetadata
+        }
+
+        return acc // This will never happen, a Promise is either rejected or fullfilled
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts:148]: Remove comment

File: src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts [hunk 3]
```diff
@@ -26,9 +39,6 @@ export const cosmosSdkOpportunityIdsResolver = async ({
 
   const chainAdapters = getChainAdapterManager()
   const portfolioAccountIds = selectWalletAccountIds(state)
-
-  // Not AccountIds of all Cosmos SDK chains but only a subset of current and future Cosmos SDK chains we support/may support
-  // We can't just check the chainNamespace, since this includes Thorchain and possibly future chains which don't use the regular Cosmos SDK staking module
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts:31]: Revert

File: src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts (hunk 1 above)
Comment (0xdef1cafe) [reply src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts:87]: this business logic around default validators has always been incredibly confusing to me at least, let's chat and confirm requirements with @DiggyDiggy2 @reallybeard and maybe even write a (gasp) test for it with this fresh implementation

File: src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        const asset = selectAssetById(state, assetId)
+        if (!asset) throw new Error(`No asset found for AssetId: ${assetId}`)
+        const marketData = selectMarketDataById(state, assetId)
+
+        const underlyingAssetRatioBaseUnit = bn(1).times(bn(10).pow(asset.precision)).toString()
+        return {
+          validatorId,
+          apy: data.apr,
+          tvl: bnOrZero(data.tokens)
+            .div(`1e+${asset.precision}`)
```
Comment (0xdef1cafe) [inline src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts:118]: pow pow

File: src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts (hunk 1 above)
Comment (gomesalexandre) [reply src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts:87]: Sounds good to me, safer with thorough tests and less fluff in this PR as well if we PR it in isolation - I'll add it as a follow-up after we confirm requirements

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**